### ISSUE TITLE
cuda memory leak with python api?

### ISSUE THREAD

**Issue by mali-nuist** [OpenNMT/CTranslate2#58]
I tranied a model which size is about 460M. The cuda memory was allocated about 669M to this model when it was loaded into python environment :
``
     import ctranslate2 
``
``
  translator = ctranslate2.Translator("/data/ende_ctranslate2/", device="cuda")
``             
My first quesiton is why the loaded model occupyied much more memory than the model size? 

When i tried to translate the frist one batch of sentence:
``
                     translator.translate_batch([["▁H", "ello", "▁world", "!"]])
``
The cuda memory occpuied by this model gradually increaed but suddenly reached about 2600M and quickly falled to 800M finally.  I quiet want know what happend during this peorid as this behavior always lead to my other programs running on the smme gpu cuda out memory error .
 

Besides, when i translate some longer sentences, the memory occpuied by this model will always increase and never decrease to the previous size.  This is quiet abnormal and i wonder whether these phenomens are resulted by memory leak?   Thanks.

**Comment by guillaumekln** [comment]
> My first quesiton is why the loaded model occupyied much more memory than the model size?

There could be multiple reasons to this (I can't put exact numbers unfortunately):
* Some space is used by CUDA internals
* Memory usage is usually larger than what was actually allocated because of padding and alignment (which are done for performance reasons).

> I quiet want know what happend during this peorid as this behavior always lead to my other programs running on the smme gpu cuda out memory error .

Running a translation requires to allocate multiple buffers for intermediate results. This could quickly add up. We also use a [caching allocator](https://nvlabs.github.io/cub/structcub_1_1_caching_device_allocator.html) that don't always free the memory. This is important for performance.

> This is quiet abnormal and i wonder whether these phenomens are resulted by memory leak?

This does not look like a memory leak, just a variable memory based on the usage. We could expose the configuration of the caching allocator to tradeoff speed and memory usage.

Overall, CTransate2 should still use less memory than a typical PyTorch or TensorFlow implementation, especially in idle state.

**Comment by mali-nuist** [comment]
@guillaumekln Thanks for you detailed replies. 

We normally start multi gpu application instances on one gpu to fully utilizing the computation power of gpu. But as we could not figure out how much memory the program will actually take under this condition, should we only start one instance on one gpu? Will the one-app on one gpu setup waste a lot of computation power?  
Or If we run multi-instances on one gpu, how much free memory should i keep for the cache allocator buffer to avoid cuda out memory error?

Thanks.

**Comment by guillaumekln** [comment]
Currently there is no guarantee regarding the maximal memory the engine can use. You could probably find the upper bound experimentally if you know the maximum request size (largest batch size and sentence length) and then distribute your instances accordingly.

Of course we are always looking to reduce the memory usage and improve the caching mechanism so this could change in future releases.

**Comment by gxzks** [comment]
@guillaumekln I also met the same issue.
When I load the model, gpu memory usage increased to be 983M. After I used `translate_batch` to translate a batch with only one sentence,  gpu memory usage increased to be 1373M but didn't fall back. When I double the sentence, it increased to be 1377M. But when I used `del translator` to release the translator, gpu memory usage fell back to be 1047M not the initial usage 983M. What is the main reason causing such phenomena?

**Comment by guillaumekln** [comment]
I'm reviewing the GPU memory management. When destroying a translator, there are 2 types of leftovers:

1. memory reserved by cuBLAS, cuDNN, and TensorRT libraries
2. cached memory for fast reuse by future translators

For 1., the PR #59 should delete some of this memory each time a specific thread terminates (this could actually be considered a memory leak if threads are never be reused again).
For 2. I could expose a `clear_cache()` function if that is useful to you.

**Comment by guillaumekln** [comment]
Also #60 will make the CUDA caching allocator configurable so that expert users can trade-off memory usage and speed. The current defaults mostly prioritize speed.

**Comment by mali-nuist** [comment]
@guillaumekln Thanks, we will try the ways you suggested and hope for better solution in the future.
